### PR TITLE
[MRG] checks that the threshold is not higher than number of maps for canICA 

### PR DESCRIPTION
**Pull request by SylvainLan**
#721

### CODE REVIEW

File: nilearn/decomposition/canica.py [hunk 1]
```diff
@@ -127,13 +127,18 @@ def __init__(self, mask=None, n_components=20, smoothing_fwhm=6,
             mask_strategy=mask_strategy, mask_args=mask_args,
             memory=memory, memory_level=memory_level,
             n_jobs=n_jobs, verbose=verbose)
+
+        if(threshold > n_components):
```
Comment (KamalakerDadi) [inline nilearn/decomposition/canica.py]: From the doc, the given threshold can be None, 'auto' or float. Here, you should also check that the given threshold should be a ```numbers.Number``` to raise an error. Also, in Python I think we don't need parenthesis.

File: nilearn/decomposition/canica.py [hunk 2]
```diff
@@ -127,13 +127,18 @@ def __init__(self, mask=None, n_components=20, smoothing_fwhm=6,
...
+            raise ValueError("Threshold must not be higher than number"
```
Comment (KamalakerDadi) [inline nilearn/decomposition/canica.py]: In the error message, there is no space after number. You can leave space after number before ".
Error message can be seen here, https://travis-ci.org/nilearn/nilearn/jobs/218779100#L2729

File: nilearn/decomposition/canica.py [hunk 3]
```diff
@@ -127,13 +127,18 @@ def __init__(self, mask=None, n_components=20, smoothing_fwhm=6,
...
+                             "of maps."
+                             "Number of maps is %s and you provided %s" %
+                             (str(n_components), str(threshold)))
         self.threshold = threshold
         self.n_init = n_init
 
     def _unmix_components(self):
         """Core function of CanICA than rotate components_ to maximize
         independance"""
-
```
Comment (KamalakerDadi) [inline nilearn/decomposition/canica.py]: Please don't remove lines which are not related to your changes.

**Comment by KamalakerDadi** [comment]
Also, you forgot to add bounds in the ```threshold``` parameter documentation. From the issue, if I get correctly threshold should be between [0. and n_components]

**Comment by KamalakerDadi** [comment]
You should also add tests to check whether error is raised or not. Tests should be added in nilearn/decomposition/tests/test_canica.py

**Comment by KamalakerDadi** [comment]
>Also, you forgot to add bounds in the threshold parameter documentation. From the issue, if I get correctly 
threshold should be between [0. and n_components]

May be it would be good idea to put threshold bound [0. and n_components] near the description of float value. Right after non-zero voxels). .......

**Comment by KamalakerDadi** [comment]
>You should also add tests to check whether error is raised or not. Tests should be added in nilearn/decomposition/tests/test_canica.py

Could you add tests please ?

File: nilearn/decomposition/tests/test_canica.py [hunk 4]
```diff
@@ -135,3 +135,12 @@ def test_component_sign():
                 canica.components_)):
             mp = mp.get_data()
             assert_less_equal(-mp.min(), mp.max())
+
+
+def test_threshold_bound():
+    data, mask_img, components, rng = _make_canica_test_data()
```
Comment (KamalakerDadi) [inline nilearn/decomposition/tests/test_canica.py]: We don't need this because directly validating the parameters where data is not required.

File: nilearn/decomposition/tests/test_canica.py [hunk 5]
```diff
@@ -135,3 +135,12 @@ def test_component_sign():
...
+
+    # Smoke test to make sure an error is raised when threshold
+    # is higher than number of components
+    assert_raises(ValueError, canICA, n_components=4, random_state=rng,
```
Comment (KamalakerDadi) [inline nilearn/decomposition/tests/test_canica.py]: Should be doing like this. Below CanICA is an object followed by inputs
```python
 assert_raises(ValueError, CanICA, n_components=4, threshold=5.)
```

File: nilearn/decomposition/canica.py [hunk 6]
```diff
@@ -127,13 +128,18 @@ def __init__(self, mask=None, n_components=20, smoothing_fwhm=6,
             mask_strategy=mask_strategy, mask_args=mask_args,
             memory=memory, memory_level=memory_level,
             n_jobs=n_jobs, verbose=verbose)
+
+        if isinstance(threshold, float) and threshold > n_components:
+            raise ValueError("Threshold must not be higher than number "
+                             "of maps. "
+                             "Number of maps is %s and you provided %s" %
```
Comment (KamalakerDadi) [inline nilearn/decomposition/canica.py]: Can you include threshold in the error message. I think writing as and "you provided threshold=%s" would be bit clear.

File: nilearn/decomposition/tests/test_canica.py [hunk 7]
```diff
@@ -135,3 +135,9 @@ def test_component_sign():
                 canica.components_)):
             mp = mp.get_data()
             assert_less_equal(-mp.min(), mp.max())
+
+
+def test_threshold_bound():
+    # Smoke test to make sure an error is raised when threshold
+    # is higher than number of components
+    assert_raises(ValueError, canICA, n_components=4, threshold=5.)
```
Comment (KamalakerDadi) [inline nilearn/decomposition/tests/test_canica.py]: It should 'C' capital (CanICA).

For reference ```from nilearn.decomposition import CanICA```.

**Comment by KamalakerDadi** [comment]
LGTM

**Comment by bthirion** [comment]
LGTM

**Comment by GaelVaroquaux** [comment]
LGTM. Merging
